specs/core/0_beacon-chain.md

```diff
@@ -1023,7 +1027,7 @@ First, some helper functions:
 ```python
 def min_empty_validator_index(validators: List[ValidatorRecord], current_slot: int) -> int:
     for i, v in enumerate(validators):
-        if v.balance == 0 and v.latest_status_change_slot + ZERO_BALANCE_VALIDATOR_TTL <= current_slot:
+        if v.balance == WITHDRAWN and v.latest_status_change_slot + ZERO_BALANCE_VALIDATOR_TTL <= current_slot:
```

Comment:
I think you meant `v.status == WITHDRAWN` but:

1) We don't have a `WITHDRAWN` status
2) The `EXITED_WITHOUT_PENALTY` and `EXITED_WITH_PENALTY` are the "withdrawable" statuses, so that combined with `v.balance == 0` is the safest/cleanest IMO. (It's safest because it protects against topups being processed after an exit, which can happen because there's a fundamental race condition.)

Reply:
hmmm... I'll leave as is for now.
I think that we need to add `WITHDRAWN` back in to remedy some of the Penalty problems, but I'll address that in another PR.

Are you suggesting that the following should be able to occur?
* validator withdraws
* 1 ETH deposit is made to this withdrawn validator
* validator can withdraw the 1 eth balance

Reply:
I'm thinking we are going to need some sort of distinction between `EXITED_WITH_PENALTY` and actually `EXITED_AND_PENALIZED` once the penalty is incurred. 

Right now penalties are broken

Reply:
> hmmm... I'll leave as is for now

If so you probably want to define `WITHDRAWN` in the constants and change `v.balance` to `v.status`.

> Are you suggesting that the following should be able to occur?

Yes. Checking `v.balance == 0` is asynchronously safe, but `v.status == WITHDRAWN` is not (because of the race condition).